### PR TITLE
fix custom plugin monitor condition change

### LINKED ISSUE #202: Reason change once and it stay in wrong state with custom plugins
Hello,

I found something again I think but maybe I follwed wrong way. 
I'm using a custom config based on the ntp example: 

```
{
    "plugin": "custom",
    "pluginConfig": {
        "invoke_interval": "30s",
        "timeout": "5s",
        "max_output_length": 80,
        "concurrency": 3
    },
    "source": "ntp-custom-plugin-monitor",
    "conditions": [
        {
            "type": "CustomProblem",
            "reason": "CustomIsUp",
            "message": "Status of the custom service"
        }
    ],
    "rules": [
        {
            "type": "permanent",
            "condition": "CustomProblem",
            "reason": "CustomIsDown",
            "path": "/usr/bin/custom.sh",
            "timeout": "3s"
        }
    ]
}

```
The /usr/bin/custom.sh script is very simple: exit with 0 or 1. 

So when the node problem detector start it set the condition:

```
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  CustomProblem    False   Wed, 05 Sep 2018 14:40:26 +0200   Wed, 05 Sep 2018 14:40:25 +0200   CustomIsUp                   Status of the custom service
  OutOfDisk        False   Wed, 05 Sep 2018 14:40:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure   False   Wed, 05 Sep 2018 14:40:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Wed, 05 Sep 2018 14:40:23 +0200   Wed, 05 Sep 2018 13:17:06 +0200   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Wed, 05 Sep 2018 14:40:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Wed, 05 Sep 2018 14:40:23 +0200   Thu, 30 Aug 2018 17:35:04 +0200   KubeletReady                 kubelet is posting ready status

```
After it run the script (what returned with 0 in this case) the Status stay false but the Reason field changed to what I set in the rule section:

```
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  CustomProblem    False   Wed, 05 Sep 2018 14:41:56 +0200   Wed, 05 Sep 2018 14:40:55 +0200   CustomIsDown                 Status of the custom service
  OutOfDisk        False   Wed, 05 Sep 2018 14:42:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure   False   Wed, 05 Sep 2018 14:42:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Wed, 05 Sep 2018 14:42:23 +0200   Wed, 05 Sep 2018 13:17:06 +0200   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Wed, 05 Sep 2018 14:42:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Wed, 05 Sep 2018 14:42:23 +0200   Thu, 30 Aug 2018 17:35:04 +0200   KubeletReady                 kubelet is posting ready status
```

So ok, in the next run the script exited with 1. The Status is True, and the Reason still same (this is what I set under the rule):

```
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  CustomProblem    True    Wed, 05 Sep 2018 14:43:56 +0200   Wed, 05 Sep 2018 14:43:55 +0200   CustomIsDown                 Status of the custom service
  OutOfDisk        False   Wed, 05 Sep 2018 14:43:53 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure   False   Wed, 05 Sep 2018 14:43:53 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Wed, 05 Sep 2018 14:43:53 +0200   Wed, 05 Sep 2018 13:17:06 +0200   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Wed, 05 Sep 2018 14:43:53 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Wed, 05 Sep 2018 14:43:53 +0200   Thu, 30 Aug 2018 17:35:04 +0200   KubeletReady                 kubelet is posting ready status
```

In the next round the script returned with 0 again and Status changed back to false but the Reason didn't change: 

```
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  CustomProblem    False   Wed, 05 Sep 2018 14:44:26 +0200   Wed, 05 Sep 2018 14:44:25 +0200   CustomIsDown                 Status of the custom service
  OutOfDisk        False   Wed, 05 Sep 2018 14:44:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure   False   Wed, 05 Sep 2018 14:44:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Wed, 05 Sep 2018 14:44:23 +0200   Wed, 05 Sep 2018 13:17:06 +0200   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Wed, 05 Sep 2018 14:44:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Wed, 05 Sep 2018 14:44:23 +0200   Thu, 30 Aug 2018 17:35:04 +0200   KubeletReady                 kubelet is posting ready status
```

As I see you overwrite the condition's rule and maybe the original condition lost and the node problem detector never can't set it again. [https://github.com/kubernetes/node-problem-detector/blob/master/pkg/custompluginmonitor/custom_plugin_monitor.go#L140](https://github.com/kubernetes/node-problem-detector/blob/master/pkg/custompluginmonitor/custom_plugin_monitor.go#L140)
But maybe I missed something.

Thank you!

### PR DESCRIPTION
**Pull request by andyxning**
Fix #202 

/cc @aga20

### CODE REVIEW

File: pkg/custompluginmonitor/custom_plugin_monitor.go [hunk 1]
```diff
@@ -126,7 +126,8 @@ func (c *customPluginMonitor) generateStatus(result cpmtypes.Result) *types.Stat
 			condition := &c.conditions[i]
 			if condition.Type == result.Rule.Condition {
 				status := toConditionStatus(result.ExitStatus)
-				if condition.Status != status || condition.Reason != result.Rule.Reason {
+				// change 1: Condition status change from True to False/Unknown
+				if condition.Status != status && condition.Status == types.True {
```
Comment (Random-Liu) [inline pkg/custompluginmonitor/custom_plugin_monitor.go]: `condition.Status == types.True && status != types.True`

File: pkg/custompluginmonitor/custom_plugin_monitor.go [hunk 2]
```diff
@@ -135,9 +136,58 @@ func (c *customPluginMonitor) generateStatus(result cpmtypes.Result) *types.Stat
 						result.Rule.Reason,
 						timestamp,
 					))
+
+					condition.Status = status
+					condition.Reason = result.Rule.Reason
+				} else if condition.Status != status && status == types.True {
```
Comment (Random-Liu) [inline pkg/custompluginmonitor/custom_plugin_monitor.go]: condition.Status != types.True && status == types.True

File: pkg/custompluginmonitor/custom_plugin_monitor.go (hunk 1 above)
Comment (Random-Liu) [reply pkg/custompluginmonitor/custom_plugin_monitor.go]: IIRC, `true` means there is a problem, `false` means there is no problem. If so, we should reset default here.

File: pkg/custompluginmonitor/custom_plugin_monitor.go [hunk 3]
```diff
@@ -135,9 +136,58 @@ func (c *customPluginMonitor) generateStatus(result cpmtypes.Result) *types.Stat
...
+					condition.Transition = timestamp
+					condition.Message = result.Message
+					events = append(events, util.GenerateConditionChangeEvent(
+						condition.Type,
+						status,
+						condition.Reason,
+						timestamp,
+					))
 				}
-				condition.Status = status
```
Comment (Random-Liu) [inline pkg/custompluginmonitor/custom_plugin_monitor.go]: I feel like most code in the if-else clause can be extracted out.

At least event generation can be extracted.

File: pkg/custompluginmonitor/custom_plugin_monitor.go (hunk 2 above)
Comment (andyxning) [reply pkg/custompluginmonitor/custom_plugin_monitor.go]: Done

File: pkg/custompluginmonitor/custom_plugin_monitor.go (hunk 1 above)
Comment (andyxning) [reply pkg/custompluginmonitor/custom_plugin_monitor.go]: > What about this? I feel like the current logic in your change is just opposite.

In your code:
status true -> false: reason is updated;
status false -> true: reason is reset to default.

Is that correct?

After a re-think, i am wrong. I will update this.

**Comment by aga20** [comment]
Seems ok with manual testing!

File: pkg/custompluginmonitor/custom_plugin_monitor.go [hunk 4]
```diff
@@ -135,9 +173,21 @@ func (c *customPluginMonitor) generateStatus(result cpmtypes.Result) *types.Stat
 						result.Rule.Reason,
 						timestamp,
 					))
+
+					condition.Status = status
+					condition.Reason = result.Rule.Reason
+				} else if condition.Status == status && condition.Message != result.Message {
```
Comment (Random-Liu) [inline pkg/custompluginmonitor/custom_plugin_monitor.go]: WE should only update condition when reason changes.

Message could be spammy and change a lot. See https://github.com/kubernetes/node-problem-detector/blob/master/pkg/systemlogmonitor/log_monitor.go#L142

Comment (Random-Liu) [reply pkg/custompluginmonitor/custom_plugin_monitor.go]: @andyxning Address this?

**Comment by Random-Liu** [comment]
I'll merged this now and we'll send a fix for https://github.com/kubernetes/node-problem-detector/pull/203#discussion_r234846858.

**Pull request merged**